rfd/0028-cluster-config-resources.md

```diff
@@ -0,0 +1,176 @@
+    // server startup.
+    string ClusterID;
+
+    // SessionRecording controls where (or if) the session is recorded.
+    string SessionRecording;
+
+    // ProxyChecksHostKeys is used to control if the proxy will check host keys
+    // when in recording mode.
+    string ProxyChecksHostKeys;
+
+    // Audit is a section with audit config
+    AuditConfig Audit;
+
+    // ClientIdleTimeout sets global cluster default setting for client idle
+    // timeouts
+    int64 ClientIdleTimeout;
+
+    // DisconnectExpiredCert provides disconnect expired certificate setting -
+    // if true, connections with expired client certificates will get disconnected
+    bool DisconnectExpiredCert;
+
+    // KeepAliveInterval is the interval the server sends keep-alive messsages
+    // to the client at.
+    int64 KeepAliveInterval;
+
+    // KeepAliveCountMax is the number of keep-alive messages that can be missed
+    // before
+    // the server disconnects the connection to the client.
+    int64 KeepAliveCountMax;
+
+    // LocalAuth is true if local authentication is enabled.
+    bool LocalAuth;
+
+    // SessionControlTimeout is the session control lease expiry and defines
+    // the upper limit of how long a node may be out of contact with the auth
+    // server before it begins terminating controlled sessions.
+    int64 SessionControlTimeout;
+}
+```
+
+It is proposed to distribute them in the following fashion:
+
+```proto
+message ClusterMeta {
+    // This resource is not genuine configuration, it would contain only the
+    // common resource metadata.  The metadata name field would be used to
+    // represent the ClusterID.
+}
+
+message SessionRecordingConfig {
+    string SessionRecording;
+    string ProxyChecksHostKeys;
+}
+
+message ClusterNetworkingConfig {
+    int64 ClientIdleTimeout;
+    int64 KeepAliveInterval;
+    int64 KeepAliveCountMax;
+    int64 SessionControlTimeout;
+}
+
+// The already existing AuditConfig is turned into a standalone resource.
+message AuditConfig { ... }
+
+// LocalAuth field is dropped altogether.  This information should be inferred
+// from ClusterAuthPreference.Type.
+
+// DisconnectExpiredCert field is moved into ClusterAuthPreference.
+message ClusterAuthPreference {
+    ...
+    bool DisconnectExpiredCert;
+}
+```
+
+### Data migration
+
+A new step should be added to `migrateLegacyResources` (called during auth
+server initialization) that takes care of migrating the `ClusterConfig` data
+stored in the backend to the new resource structures.
+
+### Restricting to a subset of values of a field
+
+Some users should not be capable of setting a field to any of the otherwise
+accepted values.  It is proposed to satisfy this requirement by introducing new
+RBAC verbs, interpreted as implying additional constraints. Such new verbs
+would be evaluated together with other RBAC verbs in the auth API handlers.
+
+In particular, this is needed for `SessionRecordingConfig` which has two
```

Comment:
May be call it session config in case we add any other session related parameters?

Reply:
I considered it but it seemed too vague and potentially confusing given that the term "session" itself can be easily (mis)interpreted in many different ways.

For instance, would the field `SessionControlTimeout` be expected to be grouped under `SessionConfig` or `ClusterNetworkingConfig` (together with the other timeout values)?

Reply:
ok fair